TX malleability End 2 End test

steevi opened the pull request.

ejfitzgerald committed:
Add counter field to transaction

ejfitzgerald committed:
Re-enable chain unit tests

ejfitzgerald committed:
Update transaction serialisation support to v2

ejfitzgerald committed:
Add support for counter in builder

ejfitzgerald committed:
Correct transaction hash to be signature agnostic

ejfitzgerald committed:
Add tests for hashes

ejfitzgerald committed:
Reformat test file

ejfitzgerald committed:
Automatic code style

steevi committed:
Merge remote-tracking branch 'Ed/feature/tx-malleability' into feature/tx-malleability

steevi committed:
Added signature vulnerability test

steevi committed:
modified end_to_end_tet.yaml

steevi committed:
updated version for python api

steevi committed:
Removed acccess.log

steevi committed:
Uncommented tests

steevi committed:
Fixed spacing

steevi committed:
Applied styleguide

ejfitzgerald commented:
@steevi can you update this PR. We have landed the transaction fixes so the diff should be much clearer now

ejfitzgerald reviewed (changes requested):
See original comment 😁

ejfitzgerald commented on scripts/end_to_end_test/smart_contract_tests/malleability_test.py:
```diff
@@ -0,0 +1,55 @@
+    id2PrivateKey = "499f40c1bf13e7716e62431b17ab668fa2688b7c94a011a3aab595477bc68347"
+
+    # Create identities from the private keys
+    id1 = Entity().from_hex(id1PrivateKey)
```
```suggestion
    id1 = Entity.from_hex(id1PrivateKey)
```

ejfitzgerald commented on scripts/end_to_end_test/smart_contract_tests/malleability_test.py:
```diff
@@ -0,0 +1,55 @@
+
+    # Create identities from the private keys
+    id1 = Entity().from_hex(id1PrivateKey)
+    id2 = Entity().from_hex(id2PrivateKey)
```
```suggestion
    id2 = Entity.from_hex(id2PrivateKey)
```

ejfitzgerald commented on scripts/end_to_end_test/smart_contract_tests/malleability_test.py:
```diff
@@ -0,0 +1,55 @@
+    api.sync(api.tokens.wealth(id1, 1000))
+
+    # transaction payload that transfers 250 FET from id1 to id2. Signed with (r,s)
+    orig_tx = b'\xa1D\x00\x00\xc5\xab \xe3\xab\x84\\\xb4\xa1\xd2\xc3\xe4\xc3\xb0\x8f_\xf4*o\xf2\xa7\x1dv\x97\xba\x8f2\xc4\x15\xb7|\x7f\x8d\x85\x0b>\xf0%\xb1\x89\xa2\xd9\xbbJQZ\x84\xc3g=\xb6\xd3\xef%8],\x8d\x1e4\xb0n-\xe1\xc0\xfa\xc0\x85\x01\x14\x00\x00\x00\x00\x00\x00\x00\x00\x04\x0b\xaf\xbca\xa0\x85$7/I]\x9d\xee\x08\xad\xbc9\x82N\x98\x05\x06\x94p\x919\\\xec\xe1f6\xdd\xc0\x94\xb9\xd4\t\xd5\xb3N\x ...
```
For readability I think that I would encode this pre-encoded transactions as hex instead of byte literals

ejfitzgerald commented on scripts/end_to_end_test/smart_contract_tests/malleability_test.py:
```diff
@@ -0,0 +1,55 @@
+    mal_tx = b"\xa1D\x00\x00\xc5\xab \xe3\xab\x84\\\xb4\xa1\xd2\xc3\xe4\xc3\xb0\x8f_\xf4*o\xf2\xa7\x1dv\x97\xba\x8f2\xc4\x15\xb7|\x7f\x8d\x85\x0b>\xf0%\xb1\x89\xa2\xd9\xbbJQZ\x84\xc3g=\xb6\xd3\xef%8],\x8d\x1e4\xb0n-\xe1\xc0\xfa\xc0\x85\x01\x14\x00\x00\x00\x00\x00\x00\x00\x00\x04\x0b\xaf\xbca\xa0\x85$7/I]\x9d\xee\x08\xad\xbc9\x82N\x98\x05\x06\x94p\x919\\\xec\xe1f6\xdd\xc0\x94\xb9\xd4\t\xd5\xb3N\xf ...
+
+    # Creating jsons for the above mentioned transactions
+    legit_trans = api.tokens._post_tx_json(orig_tx, ENDPOINT)
```
`_post_tx_json` is not a private method on this token api object. So this would be a little brittle. There is a free function that you can use `submit_json_transaction` that is part of the API directly which might be a better option

ejfitzgerald commented on scripts/end_to_end_test/smart_contract_tests/malleability_test.py:
```diff
@@ -0,0 +1,55 @@
+
+    # Sending the transactions to the ledger
+    api.sync(legit_trans)
+    api.sync(mal_trans)
```
```suggestion
api.sync([legit_trans, mal_trans])
```
Also worth making sure that `legit_trans` and `mal_trans` are the same value

steevi committed:
Resolving PR requests

steevi committed:
Resolving PR requests

steevi committed:
Resolving PR requests

steevi committed:
Waiting for the following modifications of the submit_json_function to test that the test is working.
1. submit_json_function should return the transaction hash and not True of False
2. submit_json_function should accept an endpoint as input
3. the endpoint should be passed to the format_contract_url function and the url should be created like below
`    if endpoint is None:
         url = '{}://{}:{}/api/contract/submit'.format(protocol, host, port)
     else:
         url = '{}://{}:{}/{}'.format(protocol, host, port, endpoint)`

steevi commented:
Waiting for the following modifications of the submit_json_function to verify that the test is working.

1. submit_json_function should return the transaction hash and not True or False
2. submit_json_function should accept an endpoint as input
3. the endpoint should be passed to the format_contract_url function. If I pass `endpoint='transfer'`, the url should be `protocol://host:port/api/contract/fetch/token/transfer`

Could you please also update the fetchai-ledger-api on the pypi?

steevi committed:
1. Modified test file
2. Modified ledger-api-py dependency

Commit:
Merge branch 'master' into feature/tx-malleability

steevi committed:
Applied code style

steevi committed:
Merge branch 'feature/tx-malleability' of https://github.com/steevi/ledger into feature/tx-malleability

steevi committed:
Jenkins failed building only in Clang6. Trigger for rebuild

fetch-bot committed:
Merge branch 'master' into feature/tx-malleability
